src/Language/PureScript/Linter/Exhaustive.hs

```diff
@@ -0,0 +1,228 @@
+  concatMap (\cp -> missingCasesSingle env cp cb) allPatterns
+  where
+  allPatterns = map (\(p, t) -> ConstructorBinder (qualifyProperName p con) (initialize $ length t)) $ getConstructors env con
+missingCasesSingle env cb@(ConstructorBinder con bs) (ConstructorBinder con' bs')
+  | con == con' = map (ConstructorBinder con) (missingCasesMultiple env bs bs')
+  | otherwise = [cb]
+missingCasesSingle _ NullBinder (ArrayBinder bs)
+  | null bs = [] 
+  | otherwise = []
+missingCasesSingle env NullBinder (ObjectBinder bs) =
+  map (ObjectBinder . zip (map fst bs)) allMisses
+  where
+  allMisses = missingCasesMultiple env (initialize $ length bs) (map snd bs)
+missingCasesSingle env (ObjectBinder bs) (ObjectBinder bs') =
+  map (ObjectBinder . zip sortedNames) $ uncurry (missingCasesMultiple env) (unzip binders)
+  where
+  sortNames = sortBy (compare `on` fst)
+
+  (sbs, sbs') = (sortNames bs, sortNames bs')
+
+  compB :: a -> Maybe a -> Maybe a -> (a, a)
+  compB e b b' = (fm b, fm b')
+    where
+    fm = fromMaybe e
+
+  compBS :: Eq a => b -> a -> Maybe b -> Maybe b -> (a, (b, b))
+  compBS e s b b' = (s, compB e b b')
+
+  (sortedNames, binders) = unzip $ completeMissing (compBS NullBinder) sbs sbs'
+missingCasesSingle _ NullBinder (BooleanBinder b) = [BooleanBinder $ not b]
+missingCasesSingle _ (BooleanBinder bl) (BooleanBinder br) = [BooleanBinder $ bl == br]
+missingCasesSingle env b (PositionedBinder _ _ cb) = missingCasesSingle env b cb
+missingCasesSingle _ b _ = [b]
+
+-- |
+-- Returns the uncovered set of binders
+-- the first argument is the list of uncovered binders
+-- the second argument is a clause of a pattern matching def
+--
+missingCasesMultiple :: Environment -> [Binder] -> [Binder] -> [[Binder]]
+missingCasesMultiple env = go
+  where
+  go [] _ = []
+  go (x:xs) (y:ys)
+    | null miss = map (x :) (go xs ys)
+    | otherwise = map (: xs) miss ++ map (y :) (go xs ys)
+    where
+    miss = missingCasesSingle env x y
+  go _ _ = error "Argument lengths did not match in missingCasesMultiple."
+
+-- |
+-- Is the first list of binders covered by the second?
+--
+isExhaustive :: Environment -> [Binder] -> [Binder] -> Bool
+isExhaustive env bs bs' = null $ missingCasesMultiple env bs' bs
+
+isExhaustiveMultiple :: Environment -> [[Binder]] -> [Binder] -> Bool
+isExhaustiveMultiple env bss bs = foldl' (\acc bs' -> acc && isExhaustive env bs' bs) True bss
+
+-- |
+-- Guard handling
+--
+-- We say a guard is exhaustive iff it has an otherwise (true) expression.
+-- Example:
+-- f x | x < 0 = 0
+--     | otherwise = 1
+-- Is exhaustive, whereas `f x | x < 0` is not
+--
+-- The function below say whether or not a guard has an otherwise expression
+--
+isExhaustiveGuard :: Either [(Guard, Expr)] Expr -> Bool
+isExhaustiveGuard (Left gs) = not . null $ filter (\(g, _) -> isOtherwise g) gs
+  where
+  isOtherwise :: Expr -> Bool
+  isOtherwise (TypedValue _ (BooleanLiteral True) _) = True
```

Comment:
Does `otherwise` get inlined to `true`? If not, I think we need to look for the value `otherwise` explicitly. If it does, let's document it.